composer/callbacks/image_visualizer.py

```diff
@@ -0,0 +1,189 @@
+        if not isinstance(inputs, torch.Tensor):
+            raise NotImplementedError('Multiple input tensors not supported yet')
+        # Verify inputs is a valid shape for conversion to an image
+        if _check_for_image_format(inputs):
+            table = _make_input_images(inputs, self.num_images)
+            logger.data_batch({key: table})
+
+    def _log_segmented_inputs(self, state: State, logger: Logger, key: str):
+        inputs = state.batch_get_item(key=self.input_key)
+        targets = state.batch_get_item(key=self.target_key)
+        outputs = state.outputs
+
+        if not isinstance(inputs, torch.Tensor):
+            raise NotImplementedError('Multiple input tensors not supported yet')
+        if not isinstance(targets, torch.Tensor):
+            raise NotImplementedError('Multiple target tensors not supported yet')
+        if not isinstance(outputs, torch.Tensor):
+            raise NotImplementedError('Multiple output tensors not supported yet')
+
+        table = _make_segmentation_images(inputs, targets, outputs, self.num_images)
+        logger.data_batch({key: table})
+
+    def before_forward(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'input' and state.timestamp.get(self.interval.unit).value % self.interval.value == 0:
+            self._log_inputs(state, logger, 'Images/Train')
+
+    def eval_before_forward(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'input' and state.eval_timestamp.get(TimeUnit.BATCH).value == 0:
+            self._log_inputs(state, logger, 'Images/Eval')
+
+    def before_loss(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'segmentation' and state.timestamp.get(
+                self.interval.unit).value % self.interval.value == 0:
+            self._log_segmented_inputs(state, logger, 'Images/Train')
+
+    def eval_after_forward(self, state: State, logger: Logger):
+        assert isinstance(self.interval, Time)
+        if self.mode.lower() == 'segmentation' and state.eval_timestamp.get(TimeUnit.BATCH).value == 0:
+            self._log_segmented_inputs(state, logger, 'Images/Eval')
+
+
+def _make_input_images(inputs: torch.Tensor, num_images: int):
+    import wandb
+    if inputs.shape[0] < num_images:
+        num_images = inputs.shape[0]
+    images = inputs[0:num_images].data.cpu().permute(0, 2, 3, 1).numpy()
+    table = wandb.Table(columns=['Image'])
+    for image in images:
+        img = wandb.Image(image)
+        table.add_data(img)
+    return table
+
+
+def _make_segmentation_images(inputs: torch.Tensor, targets: torch.Tensor, outputs: torch.Tensor, num_images: int):
+    import wandb
+    if min([inputs.shape[0], targets.shape[0], outputs.shape[0]]) < num_images:
+        num_images = min([inputs.shape[0], targets.shape[0], outputs.shape[0]])
```

Comment:
Shouldn't `inputs.shape[0] == targets.shape[0] == outputs.shape[0]` all the time?

Reply:
I think there might be an edge case here if someone does odd things either downsampling or upsampling the data as part of the forward pass. But, 99.9% of the time these will probably all be the same.